### PR TITLE
Initial secrets do not change after Retry

### LINKED ISSUE #2823: Do Initial secrets change after Retry packet?
https://quicwg.org/base-drafts/draft-ietf-quic-tls.html#rfc.section.5.2 states that "Initial packets are protected with a secret derived from the Destination Connection ID field from the client’s first Initial packet of the connection." This sounds like a second Initial packet from the client (in response to a server's Retry) would still be encrypted using keys derived from the original DCID instead of a potentially new DCID. However, that section later mentions the client creating an Initial packet in response to a Retry packet and the DCID is selected by the server.

These two statements should be reconciled to clarify that either the initial keys stay the same for the entire connection, or that initial keys can change. (From the discussion in #2180, it sounds like an Initial sent in response to a Retry is still the same connection.)

I think there are 3 options for behavior:
1. The initial keys stay the same for the entire connection
2. The initial keys for encrypting/decrypting a packet are derived from the server's CID listed in that packet
3. The initial keys are derived from the client's first DCID, unless the server sends a Retry packet, after which the initial keys are re-derived from the server's new DCID.

These options are in my preference order.

My understanding for why initial keys change after a Retry is to support servers handling Retry statelessly. A server should still be able to use the original DCID for initial key derivation in this case because the ODCID is needed for transport parameters, and the retry token is sent in the clear in the Initial packet, so the server can recover it from the token.

Option 2 lets the keys change, but follows a simpler principle for what the keys are. When processing a packet for a connection, the information to encrypt or decrypt it is all self-contained - there is no additional connection state to check whether a new encrypter or decrypter is needed because an Initial packet was sent/received after a Retry packet.

One of the reasons why option 3 seems odd to me is that there are multiple ways the server's CID can change between Initial packets on the same connection, but Retry packets are the only case where we change the initial keys. (E.g. consider from a client's perspective sending a second ClientHello in response to a Retry packet has the initial keys change, but sending it in response to a HelloRetryRequest keeps the keys the same, even though the server likely changed its CID in the Initial packet where it sent the HRR.)

### PR DESCRIPTION
**Pull request by ianswett**
Alternate PR to #2870 

Fixes #2823

### CODE REVIEW

File: draft-ietf-quic-tls.md [hunk 1]
```diff
@@ -804,6 +803,9 @@ modifying the contents of packets from future versions.
 The HKDF-Expand-Label function defined in TLS 1.3 MUST be used for Initial
 packets even where the TLS versions offered do not include TLS 1.3.
 
+The secrets used for protecting Initial packets do not change during the
+connection, even after receiving a Retry.
+
 {{test-vectors-initial}} contains test vectors for the initial packet
 encryption.
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: Just down here (vv) there is a note that needs to be removed also.

**Comment by martinthomson** [comment]
I don't think that's a concern, because the server is still required to echo the ODCID and the client should check it.  That change is entered into the handshake transcript, so it can't be spoofed.

**Comment by marten-seemann** [comment]
@martinthomson This PR removes the transport parameter. Nothing is entered into the handshake transcript.

**Comment by martinthomson** [comment]
Oh now I see why. Ian removed the transport parameter after I reviewed it. In an update labelled 
"Update draft-ietf-quic-transport.md".

**Comment by ianswett** [comment]
Thanks for walking through that attack Marten, I was under the impression it was no longer necessary, but I'll revert that change and add the TP back.

File: draft-ietf-quic-tls.md [hunk 2]
```diff
@@ -804,17 +803,12 @@ modifying the contents of packets from future versions.
 The HKDF-Expand-Label function defined in TLS 1.3 MUST be used for Initial
 packets even where the TLS versions offered do not include TLS 1.3.
 
+The secrets used for protecting Initial packets do not change during the
+connection, even after receiving a Retry.
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: Is it worth adding some advice on how to implement this?  It's not entirely obvious.

```suggestion
connection, even after receiving a Retry.  A server that sends a Retry
therefore needs to either remember Initial protection keys or
save them in the Retry token.
```

**Comment by ianswett** [comment]
Thanks for the suggestion Martin, accepted.

File: draft-ietf-quic-tls.md [hunk 3]
```diff
@@ -804,17 +803,14 @@ modifying the contents of packets from future versions.
 The HKDF-Expand-Label function defined in TLS 1.3 MUST be used for Initial
 packets even where the TLS versions offered do not include TLS 1.3.
 
+The secrets used for protecting Initial packets do not change during the
+connection, even after receiving a Retry.  A server that sends a Retry
+therefore needs to either remember Initial protection keys or save them
```
Comment (DavidSchinazi) [inline draft-ietf-quic-tls.md]: Shouldn't this just say that the server already needs to have access to the original destination CID because of the `original_connection_id` transport parameter? I don't think the server needs to remember/encode initial keys on top of that

File: draft-ietf-quic-tls.md [hunk 4]
```diff
@@ -805,8 +805,8 @@ packets even where the TLS versions offered do not include TLS 1.3.
 
 The secrets used for protecting Initial packets do not change during the
 connection, even after receiving a Retry.  A server that sends a Retry
-therefore needs to either remember Initial protection keys or save them
-in the Retry token.
+therefore needs to either remember the original connection ID and Initial
+protection keys or save the original connection ID the Retry token.
```
Comment (DavidSchinazi) [inline draft-ietf-quic-tls.md]: typo: connection ID *in* the Retry token.

File: draft-ietf-quic-tls.md [hunk 5]
```diff
@@ -771,8 +771,7 @@ TLS 1.3 (see {{initial-secrets}}).
 ## Initial Secrets {#initial-secrets}
 
 Initial packets are protected with a secret derived from the Destination
-Connection ID field from the client's first Initial packet of the
-connection. Specifically:
+Connection ID field from the client's Initial packet. Specifically:
```
Comment (MikeBishop) [inline draft-ietf-quic-tls.md]: Why this change, when it removes the precision about *which* Initial packet?

Comment (ianswett) [reply draft-ietf-quic-tls.md]: I can't remember why I changed that, reverted.

File: draft-ietf-quic-tls.md [hunk 6]
```diff
@@ -804,17 +804,14 @@ modifying the contents of packets from future versions.
 The HKDF-Expand-Label function defined in TLS 1.3 MUST be used for Initial
 packets even where the TLS versions offered do not include TLS 1.3.
 
+The secrets used for protecting Initial packets do not change during the
+connection, even after receiving a Retry.  A server that sends a Retry
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]:
```suggestion
connection, even though the destination connection ID in client Initial packets
changes after receiving a Retry.  A server that sends a Retry
```

File: draft-ietf-quic-tls.md [hunk 7]
```diff
@@ -804,17 +804,14 @@ modifying the contents of packets from future versions.
...
+therefore needs to either remember the original connection ID and Initial
+protection keys or save the original connection ID in the Retry token.
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]:
```suggestion
or save the original connection ID in the Retry token.
```

File: draft-ietf-quic-tls.md [hunk 8]
```diff
@@ -804,17 +804,14 @@ modifying the contents of packets from future versions.
 The HKDF-Expand-Label function defined in TLS 1.3 MUST be used for Initial
 packets even where the TLS versions offered do not include TLS 1.3.
 
+The secrets used for protecting Initial packets do not change during the
+connection, even after receiving a Retry.  A server that sends a Retry
+therefore needs to either remember the original connection ID and Initial
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: No strict need to keep Initial (packet) protection keys.

```suggestion
therefore needs to either remember the original connection ID
```

**Comment by MikeBishop** [comment]
#2926 implies a good point:  After this change, the "Initial DCID" and the "original DCID" are always the same.  We should collapse the terms as part of this PR or in a follow-up if we take it.

File: draft-ietf-quic-tls.md (hunk 3 above)
Comment (ianswett) [reply draft-ietf-quic-tls.md]: Good point, MT suggested some simplifications to the text and I added a sentence clarifying that this is not a new requirement, it just changes when the connection ID is used.

File: draft-ietf-quic-tls.md [hunk 9]
```diff
@@ -805,10 +805,12 @@ The HKDF-Expand-Label function defined in TLS 1.3 MUST be used for Initial
...
-connection, even though the destination connection ID in client Initial packets
-changes after receiving a Retry.  A server that sends a Retry
-therefore needs to either remember the original connection ID
-or save the original connection ID in the Retry token.
+connection, even though the destination connection ID in client Initial
+packets changes after receiving a Retry.  A server that sends a Retry
+therefore needs to either remember the original connection ID or save
+the original connection ID in the Retry token.  Because the initial
+connection ID is included in the server's transport parameters, the only
+difference is when in the packet processing this connection ID is utilized.
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: This reads like justification for a change, rather than static text.  How about:

> The initial connection ID is needed by a server to reconstruct packet protection keys and so that it can produce the correct value for the original_connection_id transport parameter.

And are we using "initial" or "original" here?  I went with original on the basis that it distinguished this from Initial packets, but it's not clear to be that that distinction is necessary, or even a good idea.  After all, this connection ID is what is driving packet protection for those packets; the linkage might be useful.

Comment (MikeBishop) [reply draft-ietf-quic-tls.md]: At the time, we were trying to avoid confusion between the _original_ DCID, which generated the keys, and the DCID in the current Initial.  It was a useful distinction in the old design.  With this change, I think that distinction is no longer useful.
